File: pt-1/368.largest-divisible-subset.java
/*
 * @lc app=leetcode id=368 lang=java
 *
 * [368] Largest Divisible Subset
 *
 * Sort numbers and run longest-chain DP: dp[i] is the largest divisible subset
 * ending at nums[i]. A parent array reconstructs the actual subset.
 *
 * Java note: ArrayList is used for reconstruction because appending while
 * following parent pointers is natural; Collections.reverse restores order.
 *
 * Time: O(n^2). Space: O(n).
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

// @lc code=start
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        Arrays.sort(nums);
        int n = nums.length;
        int[] dp = new int[n];
        int[] parent = new int[n];
        Arrays.fill(dp, 1);
        Arrays.fill(parent, -1);

        int bestIndex = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] % nums[j] == 0 && dp[j] + 1 > dp[i]) {
                    dp[i] = dp[j] + 1;
                    parent[i] = j;
                }
            }
            if (dp[i] > dp[bestIndex]) {
                bestIndex = i;
            }
        }

        List<Integer> answer = new ArrayList<>();
        for (int cur = bestIndex; cur != -1; cur = parent[cur]) {
            answer.add(nums[cur]);
        }
        Collections.reverse(answer);
        return answer;
    }
}
// @lc code=end

File: companies/Google/368.largest-divisible-subset.go
package leetcode

//
// @lc app=leetcode id=368 lang=golang
//
// [368] Largest Divisible Subset
//
// Notes
// Sort the values and run longest-chain DP. If nums[i] is divisible by nums[j],
// any best subset ending at j can be extended by i; parent pointers reconstruct
// the actual subset. Sorting gives transitivity, so adjacent divisibility in the
// chain implies pairwise divisibility. Time: O(n^2). Space: O(n).
//
// @lc code=start

import "sort"

func LargestDivisibleSubset368(nums []int) []int {
	sort.Ints(nums)
	n := len(nums)
	if n == 0 {
		return nil
	}

	dp := make([]int, n)
	parent := make([]int, n)
	bestIndex := 0
	for i := range nums {
		dp[i] = 1
		parent[i] = -1
		for j := 0; j < i; j++ {
			if nums[i]%nums[j] == 0 && dp[j]+1 > dp[i] {
				dp[i] = dp[j] + 1
				parent[i] = j
			}
		}
		if dp[i] > dp[bestIndex] {
			bestIndex = i
		}
	}

	answer := []int{}
	for current := bestIndex; current != -1; current = parent[current] {
		answer = append(answer, nums[current])
	}
	for left, right := 0, len(answer)-1; left < right; left, right = left+1, right-1 {
		answer[left], answer[right] = answer[right], answer[left]
	}
	return answer
}

// @lc code=end
